src/compiler/checker.ts

```diff
@@ -8844,35 +9094,58 @@ module ts {
 
         /** Check a decorator */
         function checkDecorator(node: Decorator): void {
-            let expression: Expression = node.expression;
-            let exprType = checkExpression(expression);
-
+            let signature = getResolvedSignature(node);
+            let returnType = getReturnTypeOfSignature(signature);
+            if (returnType.flags & TypeFlags.Any) {
+                return;
+            }
+            
+            let expectedReturnType: Type; 
+            let diagnosticChainHead: DiagnosticMessageChain;
+            let decoratorKind: string;
             switch (node.parent.kind) {
                 case SyntaxKind.ClassDeclaration:
                     let classSymbol = getSymbolOfNode(node.parent);
                     let classConstructorType = getTypeOfSymbol(classSymbol);
-                    let classDecoratorType = instantiateSingleCallFunctionType(getGlobalClassDecoratorType(), [classConstructorType]);
-                    checkTypeAssignableTo(exprType, classDecoratorType, node);
+                    expectedReturnType = getUnionType([classConstructorType, voidType]);
+                    decoratorKind = "class";
+                    break;
+
+                case SyntaxKind.Parameter:
+                    expectedReturnType = voidType;
+                    decoratorKind = "parameter";
                     break;
 
                 case SyntaxKind.PropertyDeclaration:
-                    checkTypeAssignableTo(exprType, getGlobalPropertyDecoratorType(), node);
+                    expectedReturnType = voidType;
+                    decoratorKind = "property";
                     break;
 
                 case SyntaxKind.MethodDeclaration:
                 case SyntaxKind.GetAccessor:
                 case SyntaxKind.SetAccessor:
                     let methodType = getTypeOfNode(node.parent);
-                    let methodDecoratorType = instantiateSingleCallFunctionType(getGlobalMethodDecoratorType(), [methodType]);
-                    checkTypeAssignableTo(exprType, methodDecoratorType, node);
-                    break;
-
-                case SyntaxKind.Parameter:
-                    checkTypeAssignableTo(exprType, getGlobalParameterDecoratorType(), node);
+                    let descriptorType = createTypedPropertyDescriptorType(methodType);
+                    expectedReturnType = getUnionType([descriptorType, voidType]);

```

Comment:
Do we actually do this anywhere else? I think @RyanCavanaugh has had some objections in using `void` in a union type.